modal-go/function.go

```diff
@@ -0,0 +1,122 @@
+package modal
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	pickle "github.com/kisielk/og-rek"
+	proto "github.com/modal-labs/libmodal/modal-go/proto/modal_proto"
+)
+
+func timeNow() float64 {
+	return float64(time.Now().UnixNano()) / 1e9
+}
+
+// function references a deployed Modal Function.
+type Function struct {
+	FunctionId string
+	ctx        context.Context
+}
+
+// FunctionLookup looks up an existing Function.
+func FunctionLookup(ctx context.Context, appName string, name string, options LookupOptions) (*Function, error) {
+	ctx = clientContext(ctx)
+
+	resp, err := client.FunctionGet(ctx, proto.FunctionGetRequest_builder{
+		AppName:         appName,
+		ObjectTag:       name,
+		Namespace:       proto.DeploymentNamespace_DEPLOYMENT_NAMESPACE_WORKSPACE,
+		EnvironmentName: environmentName(options.Environment),
+	}.Build())
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Function{FunctionId: resp.GetFunctionId(), ctx: ctx}, nil
+}
+
+// Serialize function inputs to the Python pickle format.
+func (function *Function) serialize(args []interface{}, kwargs map[string]interface{}) (bytes.Buffer, error) {
+	var inputBuffer bytes.Buffer
+
+	e := pickle.NewEncoder(&inputBuffer)
+	err := e.Encode(pickle.Tuple{
+		args,
+		kwargs,
+	})
+
+	if err != nil {
+		return bytes.Buffer{}, fmt.Errorf("error pickling data: %w", err)
+	}
+	return inputBuffer, nil
+}
+
+// Deserialize from Python pickle into Go basic types.
+func (function *Function) deserialize(buffer []byte) (interface{}, error) {
+	decoder := pickle.NewDecoder(bytes.NewReader(buffer))
+	result, err := decoder.Decode()
+	if err != nil {
+		return nil, fmt.Errorf("error unpickling data: %w", err)
+	}
+
+	return result, nil
+}
+
+// Execute a single input into a remote Function.
+func (function *Function) Remote(ctx context.Context, args []interface{}, kwargs map[string]interface{}) (interface{}, error) {
```

Comment:
I like this. Another alternative is a named arg struct like this:

```
type Args struct {
  Positional []any
  Keyword map[string]any
}
```

but comparing between the two, what you have right now looks simpler, especially if you can pass in `nil`:


```go
f.Remote([]any{1, 2, 3}, nil)
f.Remote(modal.Args{[]any{1, 2, 3}})

f.Remote(nil, map[string]any{"foo": "bar"})
f.Remote(modal.Args{Keyword: map[string]any{"foo": "bar"}})
```